service/src/DCGMIOGroup.cpp

```diff
@@ -0,0 +1,541 @@
+                            std::to_string(signal_domain_type(signal_name)),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_idx < 0 || domain_idx >= m_platform_topo.num_domain(signal_domain_type(signal_name))) {
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": domain_idx out of range.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        double result = NAN;
+        m_dcgm_device_pool.update(domain_idx);
+
+        auto it = m_signal_available.find(signal_name);
+        if (it != m_signal_available.end()) {
+            result = it->second.m_devpool_func(domain_idx);
+        }
+        else {
+    #ifdef GEOPM_DEBUG
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": Handling not defined for " +
+                            signal_name, GEOPM_ERROR_LOGIC, __FILE__, __LINE__);
+    #endif
+        }
+        return result;
+    }
+
+    // Write to the control immediately, bypassing write_batch()
+    void DCGMIOGroup::write_control(const std::string &control_name, int domain_type, int domain_idx, double setting)
+    {
+        if (!is_valid_control(control_name)) {
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": " + control_name +
+                            " not valid for DCGMIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_type != control_domain_type(control_name)) {
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": " + control_name + ": domain_type must be " +
+                            std::to_string(control_domain_type(control_name)),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (domain_idx < 0 || domain_idx >= m_platform_topo.num_domain(GEOPM_DOMAIN_BOARD)) {
+            throw Exception("DCGMIOGroup::" + std::string(__func__) + ": domain_idx out of range.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        if (control_name == "DCGM::FIELD_UPDATE_RATE") {
+            m_dcgm_device_pool.update_rate(setting*1e6);
+        }
+        else if (control_name == "DCGM::MAX_STORAGE_TIME") {
+            m_dcgm_device_pool.max_storage_time(setting);
+        }
+        else if (control_name == "DCGM::MAX_SAMPLES") {
+            m_dcgm_device_pool.max_samples(setting);
+        }
+        else {
+    #ifdef GEOPM_DEBUG
+                throw Exception("DCGMIOGroup::" + std::string(__func__) + "Handling not defined for "
+                                + control_name, GEOPM_ERROR_LOGIC, __FILE__, __LINE__);
+    #endif
+        }
+    }
+
+    // Implemented to allow an IOGroup to save platform settings before starting
+    // to adjust them
+    void DCGMIOGroup::save_control(void)
+    {
```

Comment:
Maybe add a comment here to justify the empty save/restore.

If I remember right from one of your previous explanations, these controls don't need to be explicitly saved/restored because the DGCM field group takes care of that when we create/destroy it, right?

Reply:
Good idea.  
With the most recent updates we're calling the dcgm 'unwatch' in order to restore the system to the state prior to usage of the IOGroup (i.e. disabling the dcgm geopm_fields group).  This